tfx/experimental/distributed_inference/graphdef_experiments/subgraph_partitioning/beam_pipeline.py

```diff
@@ -0,0 +1,339 @@
+
+  Before executing a remote graph, we need to prepare its inputs. We first
+  get the mapping from remote graph placeholder names to parent graph input
+  names. Then, in a copy of element, we load the remote graph inputs from
+  the parent graph.
+
+  Args:
+    pcoll: A PCollection of child graph inputs not loaded yet. Each element is
+           a dictionary from remote op names to a dictionary from tensor names
+           to tensors. Here, element[child_remote_op_name] is empty now.
+    parent_remote_op_name: The remote op name of the parent graph.
+    child_remote_op_name: The remote op name of the child graph.
+    remote_op_name_to_graph_name:
+      A mapping from remote op names to graph names.
+    graph_to_remote_op_input_name_mapping:
+      A mapping from graph names to remote op names to remote graph placeholder
+      names to parent graph input names.
+      {graph name: {remote op name: {placeholder name: input name}}}.
+
+  Returns:
+    A PCollection of inputs to the child graph. Each element is a dictionary
+    from remote op names to a dictionary from tensor names to tensors. Here,
+    element[child_remote_op_name] stores the inputs of child graph.
+  """
+  parent_graph_name = remote_op_name_to_graph_name[parent_remote_op_name]
+  name_mapping = (graph_to_remote_op_input_name_mapping
+                  [parent_graph_name][child_remote_op_name])
+
+  mapping = name_mapping.items()
+  # Calling _copy_tensor multiple times may introduce a burden, since
+  # _copy_tensor invokes a deepcopy on element.
+  for child_graph_placeholder_name, parent_graph_input_name in mapping:
+
+    step_name = ("PrepareInput[Graph_%s][Input_%s]" %
+                 (child_remote_op_name, child_graph_placeholder_name))
+    pcoll = pcoll | step_name >> beam.Map(
+        _copy_tensor,
+        parent_remote_op_name,
+        _import_tensor_name(parent_graph_input_name),
+        child_remote_op_name,
+        _import_tensor_name(child_graph_placeholder_name))
+
+  return pcoll
+
+
+def _copy_tensor(  # pylint: disable=invalid-name
+    element: Dict[Text, Dict[Text, tf.Tensor]],
+    old_graph: Text, old_tensor_name: Text,
+    new_graph: Text, new_tensor_name: Text
+) -> Dict[Text, Dict[Text, tf.Tensor]]:
+  element = copy.deepcopy(element)
+  if new_graph not in element:
+    element[new_graph] = {}
+  element[new_graph][new_tensor_name] = element[old_graph][old_tensor_name]
+  return element
+
+
+@beam.ptransform_fn
+@beam.typehints.with_input_types(Dict[Text, Dict[Text, tf.Tensor]])
+@beam.typehints.with_output_types(Dict[Text, Dict[Text, tf.Tensor]])
+def _ExtractRemoteGraphOutput(  # pylint: disable=invalid-name
+    pcoll: beam.pvalue.PCollection,
+    parent_remote_op_name: Text,
+    child_remote_op_name: Text,
+    remote_op_name_to_graph_name: Mapping[Text, Text],
+    graph_name_to_specs: Mapping[Text, List[execution_spec.ExecutionSpec]],
+) -> beam.pvalue.PCollection:
+  """Extracts remote graph output.
```

Comment:
"A PTransform that extracts remote graph output."

Reply:
Changed.